pydantic_ai_slim/pydantic_ai/ag_ui.py

```diff
@@ -0,0 +1,892 @@
+class _History:
+    """A simple history representation for AG-UI protocol."""
+
+    prompt_message_id: str  # The ID of the last user message.
+    messages: list[ModelMessage]
+    tool_calls: set[str] = field(default_factory=set)
+
+
+def _convert_history(messages: list[Message]) -> _History:
+    """Convert a AG-UI history to a PydanticAI one.
+
+    Args:
+        messages: List of AG-UI messages to convert.
+
+    Returns:
+        List of PydanticAI model messages.
+    """
+    msg: Message
+    prompt_message_id: str = ''
+    result: list[ModelMessage] = []
+    tool_calls: dict[str, str] = {}  # Tool call ID to tool name mapping.
+    for msg in messages:
+        if isinstance(msg, UserMessage):
+            prompt_message_id = msg.id
+            result.append(ModelRequest(parts=[UserPromptPart(content=msg.content)]))
+        elif isinstance(msg, AssistantMessage):
+            if msg.tool_calls:
+                for tool_call in msg.tool_calls:
+                    tool_calls[tool_call.id] = tool_call.function.name
+
+                result.append(
+                    ModelResponse(
+                        parts=[
+                            ToolCallPart(
+                                tool_name=tool_call.function.name,
+                                tool_call_id=tool_call.id,
+                                args=tool_call.function.arguments,
+                            )
+                            for tool_call in msg.tool_calls
+                        ]
+                    )
+                )
+
+            if msg.content:
+                result.append(ModelResponse(parts=[TextPart(content=msg.content)]))
+        elif isinstance(msg, SystemMessage):
+            # TODO(steve): Should we handle as instructions instead of system prompt?
+            result.append(ModelRequest(parts=[SystemPromptPart(content=msg.content)]))
+        elif isinstance(msg, ToolMessage):
+            result.append(
+                ModelRequest(
+                    parts=[
+                        ToolReturnPart(
+                            tool_name=tool_calls.get(msg.tool_call_id, 'unknown'),
+                            content=msg.content,
+                            tool_call_id=msg.tool_call_id,
+                        )
+                    ]
+                )
+            )
+        elif isinstance(msg, DeveloperMessage):  # pragma: no branch
+            # TODO(steve): Should these be handled differently?
+            result.append(ModelRequest(parts=[SystemPromptPart(content=msg.content)]))
+
+    return _History(
+        prompt_message_id=prompt_message_id,
+        messages=result,
+        tool_calls=set(tool_calls.keys()),
+    )
+
+
+__all__ = [
+    'Adapter',
+    'SSE_CONTENT_TYPE',
+    'StateDeps',
+    'StateHandler',
+    'FastAGUI',
+    'agent_to_ag_ui',
+]
+
+
+# Enums.
+# TODO(steve): Remove this and all uses once https://github.com/ag-ui-protocol/ag-ui/pull/49 is merged.
+class Role(str, Enum):
```

Comment:
I think we can drop this and just hard-code string values, they're defined as Literals on the event classes so they're type-checked for validity anyway

Reply:
We could but that results in repeated values, which is less efficient. I was hoping we could get these eliminated before we merge.

Reply:
If we get them eliminated even better, but if not we prefer literals as @Viicos explained here: https://github.com/pydantic/pydantic-ai/pull/2024#discussion_r2192033809